// @react-native-assets/slider/dist/RangeSlider.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const RN = __importStar(require("react-native"));
const useRange_1 = __importDefault(require("./hooks/useRange"));
const Track_1 = __importDefault(require("./components/Track"));
const Thumb_1 = __importDefault(require("./components/Thumb"));
const ResponderView_1 = __importDefault(require("./components/ResponderView"));
const useDrag_1 = __importDefault(require("./hooks/useDrag"));
const useCustomMarks_1 = __importDefault(require("./hooks/useCustomMarks"));
// We add a default padding to ensure that the responder view has enough space to recognize the touches
const styleSheet = RN.StyleSheet.create({
    vertical: {
        paddingHorizontal: 10
    },
    horizontal: {
        paddingVertical: 10
    }
});
const Slider = react_1.default.forwardRef((props, forwardedRef) => {
    const defaultValue = react_1.default.useMemo(() => [props.minimumValue || 0, props.minimumValue || 0], [props.minimumValue]);
    const { minimumValue = 0, maximumValue = 1, range: propValue = defaultValue, step = 0, outboundColor = 'grey', inboundColor = 'blue', thumbTintColor = 'darkcyan', thumbStyle, trackStyle, minTrackStyle, midTrackStyle, maxTrackStyle, inverted = false, vertical = false, enabled = true, slideOnTap = true, trackHeight = 4, thumbSize = 15, thumbImage, minimumRange = step, crossingAllowed = false, onValueChange, onSlidingStart, onSlidingComplete, CustomThumb, CustomMark, ...others } = props;
    const { updateClosestValue, updateMaxValue, range, canMove } = (0, useRange_1.default)({
        minimumRange,
        minimumValue,
        maximumValue,
        range: propValue,
        step,
        slideOnTap,
        crossingAllowed,
        onValueChange
    });
    const { onPress, onMove, onRelease } = (0, useDrag_1.default)({ value: range, updateValue: updateClosestValue, onSlidingComplete, onSlidingStart, canMove });
    const [min, max] = range;
    const minTrackPct = react_1.default.useMemo(() => (min - minimumValue) / ((maximumValue - minimumValue) || 1), [min, minimumValue, maximumValue]);
    const maxTrackPct = react_1.default.useMemo(() => (max - minimumValue) / ((maximumValue - minimumValue) || 1), [max, minimumValue, maximumValue]);
    const { minStyle, midStyle, maxStyle } = react_1.default.useMemo(() => ({
        minStyle: (trackStyle && minTrackStyle) ? [trackStyle, minTrackStyle] : trackStyle || minTrackStyle,
        midStyle: (trackStyle && midTrackStyle) ? [trackStyle, midTrackStyle] : trackStyle || midTrackStyle,
        maxStyle: (trackStyle && maxTrackStyle) ? [trackStyle, maxTrackStyle] : trackStyle || maxTrackStyle
    }), [trackStyle, minTrackStyle, midTrackStyle, maxTrackStyle]);
    const thumbProps = react_1.default.useMemo(() => ({
        color: thumbTintColor,
        style: thumbStyle,
        size: thumbSize,
        CustomThumb: CustomThumb,
        thumbImage
    }), [CustomThumb, thumbImage, thumbSize, thumbStyle, thumbTintColor]);
    const { marks, onLayoutUpdateMarks } = (0, useCustomMarks_1.default)(CustomMark, { step, minimumValue, maximumValue, activeValues: range, inverted, vertical });
    return (react_1.default.createElement(RN.View, { ...others },
        react_1.default.createElement(ResponderView_1.default, { style: styleSheet[vertical ? 'vertical' : 'horizontal'], ref: forwardedRef, maximumValue: maximumValue, minimumValue: minimumValue, step: step, value: max, updateValue: updateMaxValue, onPress: onPress, onMove: onMove, onRelease: onRelease, enabled: enabled, vertical: vertical, inverted: inverted, onLayout: onLayoutUpdateMarks },
            react_1.default.createElement(Track_1.default, { color: outboundColor, style: minStyle, length: minTrackPct * 100, vertical: vertical, thickness: trackHeight }),
            react_1.default.createElement(Thumb_1.default, { ...thumbProps, value: min, thumb: 'min' }),
            react_1.default.createElement(Track_1.default, { color: inboundColor, style: midStyle, length: (maxTrackPct - minTrackPct) * 100, vertical: vertical, thickness: trackHeight }),
            react_1.default.createElement(Thumb_1.default, { ...thumbProps, value: max, thumb: 'max' }),
            react_1.default.createElement(Track_1.default, { color: outboundColor, style: maxStyle, length: (1 - maxTrackPct) * 100, vertical: vertical, thickness: trackHeight }),
            marks)));
});
Slider.displayName = 'Slider';
exports.default = Slider;
